Outline the type inference algorithm (#384)
ferrocene/specification: src/types-and-traits.rst

  of the :t:`constant parameter`.

* :dp:`fls_5d4hw3gj4w4n`
  The :t:`expected type` of the :t:`expression` of a :t:`constant initializer`
  is the :t:`type` specified by the :t:`type ascription` of the containing
  :t:`constant`.

* :dp:`fls_qlovdtcj1v1b`
  The :t:`expected type` of the :t:`expression` of a :t:`static initializer` is
  the :t:`type` specified by the :t:`type ascription` of the containing
  :t:`static`.

* :dp:`fls_Z5gKFjZW5rRA`
  The :t:`expected type` of the :t:`expression` of a
  :t:`discriminant initializer` is determined as follows:

  * :dp:`fls_vYvumjTQH9Xg`

>     If the :t:`enum type` containing the :t:`discriminant` is subject to

Reviewer: ```suggestion
    If the :t:`enum type` that contains the :t:`discriminant` is subject to
```

A :t:`diverging type variable` is a :t:`type variable` that can refer to any
:t:`type` and originates from a :t:`diverging expression`.

:dp:`fls_rvj3XspFZ1u3`
The :t:`type inference` algorithm uses :t:`type unification` to propagate known
:t:`[type]s` of :t:`[expression]s` and :t:`[pattern]s` across the
:t:`type inference root` being inferred. In the rules detailed below, a static
error occurs when :t:`type unification` fails.

:dp:`fls_SZgixDCAx6PQ`
:t:`Type inference` for a :t:`type inference root` proceeds as follows:

#. :dp:`fls_XYY1U9h9HlAa`
   Recursively process all :t:`[expression]s` and :t:`[statement]s` in the

>    :t:`type inference root` in program order.

Reviewer: ```suggestion
   :t:`type inference root` in declaration order.
```
We have used that term before.

Reviewer: Good point!

Author: Actually "evaluation order" is the better term here, "program order" typically means something else I believe

Reviewer: "Evaluation order" is a tricky one because in most cases it is "left-to-right", but in certain cases it is "right-to-left". Also, we have been using "declarative order" and "left-to-right order" in describing the dynamic semantics of evaluation and execution, so "evaluation order" seems to be a combination of "top-to-bottom" and "left-to-right" with the occasional "right-to-left".

   #. :dp:`fls_1rnssw39aRWn`
      For each :t:`statement`, apply the :t:`statement` inference rules below.
   
   #. :dp:`fls_aYJaZXcOVVyk`
      For each :t:`expression`, apply the :t:`expression` inference rules below.

#. :dp:`fls_X8kLC7JwiF0A`
   If there are any remaining :t:`[integer type variable]s` that have not been
   unified with a concrete :t:`integer type`, perform integer type fallback by
   unifying them with ``i32``.

#. :dp:`fls_ZFQhOxO3jpby`
   If there are any remaining :t:`[floating-point type variable]s` that have not
   been unified with a concrete :t:`floating-point type`, perform floating-point
   type fallback by unifying them with ``f64``.

#. :dp:`fls_l1G52a0qqEes`
   If there are any remaining :t:`[diverging type variable]s` that have not been

>    unified with a concrete :t:`type`, unify them with ``()``.

Reviewer: ```suggestion
   unified with a concrete :t:`type`, unify them with the :t:`unit type`.
```

#. :dp:`fls_Tx4Sx4Qy8y2d`
   If there are any remaining :t:`[global type variable]s` that have not been
   unified with a concrete :t:`type`, raise a static error.

:dp:`fls_hISRWZUuqE4Q`
The :t:`type inference` rules for :t:`[statement]s` are as follows:

* :dp:`fls_ygi1ACJ0RkfS`
  :t:`[Item statement]s` are not subject to :t:`type inference`.

* :dp:`fls_97Fxlv2KN6QF`
  :t:`[Expression statement]s` apply the :t:`expression` inference rules below

>   to the contained :t:`expression`, without any :t:`expected type` set.

Reviewer: ```suggestion
  to the related :t:`expression`, without any :t:`expected type` set.
```
